clang/lib/AST/Interp/Opcodes.td

```diff
@@ -198,11 +198,11 @@ def NoRet : Opcode {}
 
 
 def Call : Opcode {
-  let Args = [ArgFunction, ArgUint32];
+  let Args = [ArgFunction, ArgUint32, ArgExpr];
```

Comment:
Can we not get by without this? When emitting the diagnostic, the `CallExpr` should be on the caller frame, at the current frame's `RetPC`. We do something similar in `cleanupAfterFunctionCall`:
```c++
    const Expr *CallSite = S.Current->Caller->getExpr(S.Current->getRetPC());
```
or is the problem that the expr we get here might not be a `CallExpr`, but e.g. a `CXXConstructExpr`?

Reply:
Yeah, it's might be a `CXXConstructExpr` or `CXXInheritedCtorInitExpr`, for this reason, we make the `InterpFrame::CallExpr` has `Expr *` but not a `CallExpr *` type.

Reply:
As you can see in the snippet I posted, you get an `Expr*` anyway, so you can check for the type of that as well. Can you try if that works?

Reply:
Thanks for your guidance! I've try in my local, it' caused 6 test failures.
```diff
diff --git a/clang/lib/AST/Interp/InterpFrame.cpp b/clang/lib/AST/Interp/InterpFrame.cpp
index 1d285dfb5f62..36c5b586564f 100644
--- a/clang/lib/AST/Interp/InterpFrame.cpp
+++ b/clang/lib/AST/Interp/InterpFrame.cpp
@@ -172,8 +172,9 @@ void InterpFrame::describe(llvm::raw_ostream &OS) const {
       F && (F->isBuiltin() || F->isLambdaStaticInvoker()))
     return;
 
+  const auto *MyCE = S.Current->Caller->getExpr(S.Current->getRetPC());
   const FunctionDecl *F = getCallee();
-  if (const auto *MCE = dyn_cast_if_present<CXXMemberCallExpr>(CallExpr)) {
+  if (const auto *MCE = dyn_cast_if_present<CXXMemberCallExpr>(MyCE)) {
     const Expr *Object = MCE->getImplicitObjectArgument();
     Object->printPretty(OS, /*Helper=*/nullptr, S.getCtx().getPrintingPolicy(),
                         /*Indentation=*/0);
@@ -182,7 +183,7 @@ void InterpFrame::describe(llvm::raw_ostream &OS) const {
     else
       OS << ".";
   } else if (const auto *OCE =
-                 dyn_cast_if_present<CXXOperatorCallExpr>(CallExpr)) {
+                 dyn_cast_if_present<CXXOperatorCallExpr>(MyCE)) {
     OCE->getArg(0)->printPretty(OS, /*Helper=*/nullptr,
                                 S.getCtx().getPrintingPolicy(),
                                 /*Indentation=*/0);```
```console
Failed Tests (6):
  Clang :: AST/Interp/arrays.cpp
  Clang :: AST/Interp/constexpr-nqueens.cpp
  Clang :: AST/Interp/cxx20.cpp
  Clang :: AST/Interp/lambda.cpp
  Clang :: AST/Interp/records.cpp
  Clang :: SemaCXX/constexpr-frame-describe.cpp
```
```C++
constexpr int div(int a, int b) {
  auto f = [=]() {
    return a / b; // both-note {{division by zero}}
  };

  return f(); // expected-note {{in call to 'f.operator()()'}} \
              // ref-note {{in call to 'f.operator()()'}}
}
static_assert(div(8, 2) == 4);
static_assert(div(8, 0) == 4); // both-error {{not an integral constant expression}} \
                               // both-note {{in call to 'div(8, 0)'}}
```
```console
error: 'both-note' diagnostics expected but not seen: 
  File /Users/yrong/Developer/Toolchain/llvm/trunk/llvm-project/clang/test/AST/Interp/lambda.cpp Line 53: in call to 'div(8, 0)'
error: 'both-note' diagnostics seen but not expected: 
  File /Users/yrong/Developer/Toolchain/llvm/trunk/llvm-project/clang/test/AST/Interp/lambda.cpp Line 53: in call to 'f.div(8, 0)'
2 errors generated.
```
Seems it's found an incorrect CallExpr.

Reply:
You forgot the `This && IsMemberCall` check from the current interpreter.
Consider these diffs:
```diff
+
+   const Expr *MyCE = Caller->getExpr(getRetPC());
+   assert(MyCE == CallExpr);
+
   const FunctionDecl *F = getCallee();
-  if (const auto *M = dyn_cast<CXXMethodDecl>(F);
-      M && M->isInstance() && !isa<CXXConstructorDecl>(F)) {
-    print(OS, This, S.getCtx(), S.getCtx().getRecordType(M->getParent()));
-    OS << "->";
+  assert(F);
+
+  bool IsMemberCall =
+      isa<CXXMethodDecl>(F) && !isa<CXXConstructorDecl>(F) &&
+      cast<CXXMethodDecl>(F)->isImplicitObjectMemberFunction();
+
+  assert(Caller);
+
+  if (Func->hasThisPointer() && IsMemberCall) {
+    if (const auto *MCE = dyn_cast_if_present<CXXMemberCallExpr>(MyCE)) {
+      const Expr *Object = MCE->getImplicitObjectArgument();
+      Object->printPretty(OS, /*Helper=*/nullptr, S.getCtx().getPrintingPolicy(),
+                          /*Indentation=*/0);
+      if (Object->getType()->isPointerType())
+        OS << "->";
+      else
+        OS << ".";
+    } else if (const auto *OCE =
+                   dyn_cast_if_present<CXXOperatorCallExpr>(MyCE)) {
+      OCE->getArg(0)->printPretty(OS, /*Helper=*/nullptr,
+                                  S.getCtx().getPrintingPolicy(),
+                                  /*Indentation=*/0);
+      OS << ".";
+    }
```

```diff
diff --git a/clang/lib/AST/Interp/Source.cpp b/clang/lib/AST/Interp/Source.cpp
index 45cd0ad4fd42..77796b00ca52 100644
--- a/clang/lib/AST/Interp/Source.cpp
+++ b/clang/lib/AST/Interp/Source.cpp
@@ -41,7 +41,7 @@ const Expr *SourceInfo::asExpr() const {
 const Expr *SourceMapper::getExpr(const Function *F, CodePtr PC) const {
   if (const Expr *E = getSource(F, PC).asExpr())
     return E;
-  llvm::report_fatal_error("missing source expression");
+  return nullptr;
 }
```

I'll have to rebuild llvm locally now because I figured out the `llvm::report_fatal_error` above just trips my build up.

Reply:
Thanks, I've updated this PR, WDYT?

Reply:
The `assert(MyCE == CallExpr);` needs to go, but other than that, `test/AST/Interp` works. Haven't tested anything outside that yet.

Reply:
Thank you, It works!